# CRISPys
•	Background: Due to extensive history of local and large-scale genomic duplications, many eukaryotic genomes harbor homologous gene families of partially overlapping functions. This redundancy often leads to mutational robustness such that the inactivation of one gene often results in no or minimal phenotypic consequence. The CRISPR system has been recently adopted as a genome editing technique of eukaryotic genomes. The system is directed to the genomic site using a programmed single-guide RNA (sgRNA) that base-pairs with the DNA target, subsequently leading to a site-specific double-strand break, frequently leading to inactivation of the encoded protein. Yet, the low specificity of the system could, in fact, be harnessed to enable a rational design of an sgRNA that would target multiple genes simultaneously.  
•	Main advances: Previous works for accomplishing this task is to align the sequences of the given genes and then locate highly similar CRISPR-Cas9 target sites among them using the consensus sequence, while allowing for few mismatches between the consensus and each of the aligned sequences. This approach may miss the most promising sgRNA candidates. We here propose a graph-based method for finding of the most effective sgRNA candidates under alternative sets of assumptions and based the ground for farther research on choosing a group of sgRNA aim to silence large gene families. In addition, we suggested a novel approaches for transforming any topological space to a metric space, which was needed in our research.
